### ISSUE TITLE
Execute continuation in a new independent non-reentrant request

### ISSUE THREAD

**Issue by jdom** [dotnet/orleans#51]
Many times we do some work and then do an external service call (or even a grain call) that we would like to process once it returns, but in the meantime, we'd like end the current turn. Once the response comes back, it's be nice to execute the continuation as a full new turn.

Currently it's easy to do things like

```
someClient.ExecuteIoCallAsync().ContinueWith(this.MyCustomContinuation)
```

In order to end the turn but then execute the continuation... nevertheless the continuation, albeit running in a single thread in the grain, can interleave with other turns. So in order to workaround that, we sometimes have to do a grain call to tself in the callback, so that it executes as a new request and doesn't interleave. This workaround is not that trivial and still prone to errors, ignoring the fact that you have to "pollute" the public-facing grain interface with the internal details of how we deal with continuations.

Wouldn't it be nice to do something like this instead?

_UPDATE:_ In this thread the code evolved a lot, and we are currently discussing something that doesn't use _await_. I'd encourage you to look at the updated gist instead: https://gist.github.com/jdom/6597e183eabd3251f4bd

```
public async Task MyPublicGrainMethod()
{
  // do some async work in the "1st turn"
  await this.SomethingThatNeedsToRunDuringTheGrainCall();

  // now start the async call that I don't want to await during the grain call, but I want the continuation to run in a full turn.
  var myIoTask = someClient.ExecuteIoCallAsync();
  var response = await myIoTask.ConfigureAwaitInNewTurn(endCurrentTurn: true);
  // use response here in a "2nd" turn. Notice that interleaving might have happened by this time, but given you are using ConfigureAwaitInNewTurn, you should know that this is an advanced scenario and be aware of the consequences, such as using closures or ackowledging that state might have mutated by now.
}
```

I made that example verbose, so that we can discuss ideas... I don't know if it's even possible to end the turn abruptly in the middle of the method, as the task returned by MyPublicGrainMethod might not be completed yet (or can the await logic complete that task instead?, that'd be ideal).

I'm all ears and open to feedback, as this is a simplification of a scenario we use a lot.

**Comment by sergeybykov** [comment]
Maybe I have a terminology gap here. Do you mean that you want to complete the *method call* and potentially return a value before the myIoTask completes, and later process the result of myIoTask within the single-threaded guarantee but in a special context not associated with a method call?

**Comment by jdom** [comment]
Yes, I think we are talking about the same thing, but let me re-phrase it:
I'd like to finish the Orleans turn as soon as myIoTask yields (if someone was awaiting for that call to complete, it would be completed at that point). Of course this would only work with Task, but not Task<T>, as you wouldn't have a result yet.
At this point the grain can handle other calls in their own turn normally.
When myIoTask completes, then the continuation is scheduled as a new turn (meaning that it cannot be interleaved), similar to calling a grain method, only that it does not have to be exposed externally/publicly.

**Comment by sergeybykov** [comment]
Now I'm even more confused. someClient.ExecuteIoCallAsync() returns immediately. If you want to do something in the method before myIoTask completes, just don't await before you are done with. Now do you want to return from MyPublicGrainMethod() before myIoTask completes?

**Comment by ReubenBond** [comment]
You could do this with a `void Defer(Func<Task> action)` method which allows a grain to schedule an action to be executed on the next turn.

You can fake it using a one-shot timer.

**Comment by jdom** [comment]
Not technically return from the method (as that is not even a possibility in C#)... just complete the turn by then... If there is a caller waiting for that grain call to complete, then they will be done by then (as this is a proxied call, they are not really waiting on the class 'method' to complete, just on the proxied Task to complete).

This is just an idea, I could come up with some other syntax that would cover a similar scenario, but be less ambiguous, albeit more verbose, such as the following:

   public Task MyPublicGrainMethod()
   {
     // do some async work in the "1st turn"
     var myIoTask = someClient.ExecuteIoCallAsync();
     myIoTask.ContinueInNewTurn(this.MyCustomContinuation);
   }

   private Task MyCustomContinuation(Task previousTask)
   {
     var response = await previousTask;
     // use response here in a "2nd" full turn. No other turn can interleave at this point, even if this method yields. 
    // Also, this executes not as soon as myIoTask completed. Instead it got enqueued and processed normally, as if this was a normal grain call, even if it's a private method not expose to outsiders.
}

**Comment by jdom** [comment]
@ReubenBond, exactly... that was similar to what I meant with the 2nd option I gave (ContinueInNewTurn). I assume the "Defer" was a proposal, right? because I don't see that in the current Orleans API...

**Comment by gabikliot** [comment]
Sorry @jdom , but your example above does not compile. Both methods are marked to return a task but they do not. Can you please write a short but fully correct example, with numbering lines so it is clear without any doubt what execution order you would like to have.
Also, I am afraid we are confusing turns and reentrancy here. Lets be precise.
(https://github.com/dotnet/orleans/wiki/Concurrency)
(https://github.com/dotnet/orleans/wiki/Reentrant-Grains)

@ReubenBond , there is await Delay(TimeSpan) which is basically a yield point, if that is what you mean.

**Comment by jdom** [comment]
Hi @gabikliot, I've updated the code to add the async keyword to the methods so it can compile. I also created a Gist with this code so it's easier to comment on at: https://gist.github.com/jdom/6597e183eabd3251f4bd

What I'm talking about is allowing a non-reentrant grain to initiate a long running external code and then yield so the turn completes and other message processing can happen. When that long running call is done, then the continuation is executed as it's own turn, so no interleaving happens when processing that continuation.

**Comment by gabikliot** [comment]
More clarifications:
You wrote: "When that long running call is done, then the continuation is executed as it's own turn, so no interleaving happens when processing that continuation."

When continuation is executed, it is ALWAYS executed as it's own turn. Every piece of code executes as a separate turn. Perhaps what you mean is that the continuation will execute as an independent request that does not interleave with any other new requests?

**Comment by jdom** [comment]
@gabikliot ok, then I guess I had a miss-understanding about "turn", which quickly reading the link you provided didn't clarify. But yes, effectively I want the continuation to be treated as an independent request that does not interleave with other new request. Similar to doing a new grain call with that continuation, except that it's an implementation detail that is not exposed in the public grain interface.

**Comment by gabikliot** [comment]
Got you! So what you are asking for is: ContinueNonReentrant.
Or in deferent words - queue that piece of code as an independent request, subject to non-reentrancy guarantees (as opposite to "queue that piece of code as an independent turn", which is ALWAYS the case).

**Comment by jdom** [comment]
@gabikliot exactly, sorry about the confusion, I'll update the code in the gist to avoid using "turn" and replace with "request", and also update the extension method name with your suggestion ContinueNonReentrant

**Comment by gabikliot** [comment]
Ok, so lets try to see how this API may look like.
Task ContinueNonReentrant(Func<Task>);

You pass it a function that returns Task, since the logic that you want to execute inside the request may need to be async.
The return type also has to be Task, since you may want to somehow know when it was done, handle errors, ... If you don't care you can just ignore the returned Task.

Now you can't await this call. You cannot write: await ContinueNonReentrant
since it will immediately result in deadlock, as you are asking to enqueue a new request, which can only start after you are done with the currently executing request, but since you are awaiting it this request will never be done.
Similarly, you can't return this Task, as it will cause the same deadlock.
So this capability is a bit dangerous, can be easily misused. But lets put that on a side now.

If you do want to know sometime later how this request executed, you can store this task in your member field and query it in some future requests.

How does that sound?
Did I cover everything?
Do you have different suggestions?

**Comment by brinkrob** [comment]
I've been following this since it seems to be close to an issue that I've been wanting to tackle.  I think the comment from @ReubenBond was closest to what I want, but what @gabikliot proposed may work too.

Basically what I want is the ability for a grain to perform some work, return the value to the original caller, and then easily schedule a second bit of work on itself.

For Example, a web server might initiate an operation on a grain in response to a web request.  The operation might be simple and we want the response to be given back to the web server immediately so that the client is not waiting on the request.  But as a result of that request we may want to queue a longer running operation for which the end user does not need the result.

I would think it would look something like this:

    // New base member function on a Grain
    void Defer(Func<Task> continuation );

    // Custom Grain Code
    Task<int> DoSomethingQuick()
    {
        // Do quick work

        // Defer some work after this request is complete.
        Defer(UpdateLongRunningProcess);

        // finish this request by returning our answer.
        return Task.FromResult(42);
    }

    Task UpdateLongRunningProcess()
    {
        // Do something that takes a longer time
    }

If you wanted to trap errors you could have an override of the Defer method:

    void Defer(Func<Task> continuation, Func<Exception, Task> onError);

Just some ideas...

**Comment by sergeybykov** [comment]
I like the Defer() idea. And the analogy with a one-off timer is a clean one. With a response already sent to the caller, there is no way for the grain to report a failure as part of that call context, just like in the case of a timer tick. in that light, I'm not even sure the override with onError is necessary. Who's going to handle its failures?

Why not simply add Grain.Defer(Func<object, Task> continuation, object state) to mimic Grain.RegisterTimer()?

**Comment by gabikliot** [comment]
There is a way to handle errors, of course. There is no way to return this error back to the original caller in the context of the original request, true. But the error can be stored in the member field, returned later based on a later call, it cab be passed to someone else, .... It can be handled in many ways, just differently.

What is the exact signature of Grain.Defer?
Task Grain.Defer(Func<object, Task> continuation, object state)

Please notice a big difference between what @brinkrob  suggested and what @jdom suggested. @jdom asked for the deferred action to be executed in a **separate request**. @brinkrob did not ask for that. What @brinkrob asked can be achieved easily now, with just calling UpdateLongRunningProcess and not awaiting its task. Handling errors of UpdateLongRunningProcess can be done with ContinueWith.

I do think there is a lot of similarities between what @jdom and other asked I the past for proper ways to execute background jobs. So if are to add a new API, it better handle all those cases.

**Comment by jdom** [comment]
@gabikliot, I think you've got my idea, but understood something wrong about it, which might be a reason to not call it ContinueXXX to avoid the similarities with ContinueWith.

I was envisioning the ContinueNonReentrant to be of the following signature:

    public static void ContinueNonReentrant(this Task previousTask, Func<Task> nonReentrantContinuation)
    public static void ContinueNonReentrant(this Task previousTask, Func<Task, Task> nonReentrantContinuation)

And of course the generic Task equivalent for "previousTask" (not going to list all possible overloads):
    public static void ContinueNonReentrant<TResult>(this Task previousTask, Func<Task<TResult>, Task> nonReentrantContinuation)

Notice that I expect this ContinueNonReentrant extension method to have no return value (void), so it can't be awaited. That's probably a reason why NOT to call it ContinueXXX, but don't know of a better name, as ContinueXXX is very clear.
BTW, this is also kind of similar to the Defer proposal, it's just that with Defer, you immediately enqueue a request to process later, but with the ContinueNonReentrant you only enqueue a new request once "previousTask" is completed.
Of course you could use Defer to build up something similar, but it would be kind of wonky, as you'd have to await the external call first to then Defer the processing, but also call fire & forget on the original await, so that the grain call does not have to wait for the external call to complete... again, making this common scenario very complex to write.

Regarding error handling, I don't expect that to be part of the signature, you have to be clean enough to avoid unhandled exceptions in the the continuation, otherwise Orleans will just log it.

@gabikliot, also the workaround to store the task is not ideal, and it's subject to modifying state while interleaving, which can result in corruption if you are not careful enough. Also, you'd be at the mercy of the next grain call so that you can process the result, and now all of your public grain calls should have a call to process pending long running results... We have another workaround that we use today, but it's complex enough to warrant making this feature (enqueing non-reentrant continuations) a first class citizen in Orleans...

**Comment by jason-bragg** [comment]
Closing due to out of scope.
